### PR TITLE
writers: add data-centric writers

### PR DESCRIPTION
**Pull request by zimbatm**
###### Description of changes

Added some data writers that have a similar interface to the script writers. It's a simple and clean interface that makes the common case nice compared to `lib.generators`. JSON is being formatter, and YAML is actually YAML, not JSON.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Lassulus** [comment]
looks nice, seems to work. thank you!

**Review by Lassulus — APPROVED** [review]

**Comment by Lassulus** [comment]
not sure why ofborg is unhappy

**Comment by zimbatm** [comment]
Something to do with the test framework expecting derivations instead of recursive attributes. I'll take a look later.

Btw, the pypy and fsharp tests are still broken.

For pypy, it looks like pypy2.withPackages is broken.

For fsharp I don't know. I fixed the noDeps test case but there is another issue with nuget.

**Comment by zimbatm** [comment]
fixed the CI. Turns out some of the Haskell writers were duplicated ( 73ee03c )

File: pkgs/build-support/writers/data.nix [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+let
+  daselBin = lib.getExe dasel;
+
+  inherit (lib)
+    last
+    optionalString
+    types
+    ;
+in
+rec {
```
Comment (AndersonTorres) [inline pkgs/build-support/writers/data.nix:11]: Is `rec` needed?

Comment (AndersonTorres) [reply pkgs/build-support/writers/data.nix:11]: Mysterious.

**Comment by AndersonTorres** [comment]
I remember a long time ago someone tried to write a writer for a Rust-like serial object notation language.

Does this new approach help?

**Comment by lopsided98** [comment]
This breaks `writeNginxConfig` when cross-compiling, see #245785 for a fix.

File: pkgs/build-support/writers/data.nix
Comment (roberth) [inline pkgs/build-support/writers/data.nix:1]: These writers appear redundant, as we already have `pkgs.formats` for these.
If we're going to keep them, should the respective formats be changed to use these writers?

**Review by roberth — COMMENTED** [review]
This code is completely redundant and therefore a liability.
Please clean up the duplication.

**Comment by roberth** [comment]
> JSON is being formatter, and YAML is actually YAML, not JSON.

These issues have long been solved in `pkgs.formats`.

File: pkgs/build-support/writers/data.nix [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  #
+  #   input :: T -> string: function that takes the nix data and returns a string
+  #   output :: string: script that takes the $inputFile and write the result into $out
+  #   nameOrPath :: string: if the name contains a / the files gets written to a sub-folder of $out. The derivation name is the basename of this argument.
+  #   data :: T: the data that will be converted.
+  #
+  # Example:
+  #   writeJSON = makeDataWriter { input = builtins.toJSON; output = "cp $inputPath $out"; };
+  #   myConfig = writeJSON "config.json" { hello = "world"; }
+  #
```
Comment (roberth) [inline pkgs/build-support/writers/data.nix:25]: Where's the user documentation?
This is a new file, but I don't see any code to render this to the manual.

**Comment by infinisil** [comment]
Damn, I didn't know about this, indeed this pretty much duplicates `pkgs.formats`, unfortunate that this is only noticed now :disappointed:

**Comment by roberth** [comment]
Haven't heard anything in over a week.
This took 15 minutes, but 15 minutes too much nonetheless.
- https://github.com/NixOS/nixpkgs/pull/263096

Would have appreciated at least *some* response.

**Comment by Mic92** [comment]
Maybe this was drowned in below other notifications.

**Comment by roberth** [comment]
I'm sure it has. I put a lot of (too much?) effort into at least scanning for notifications where I'm responsible.

**Comment by Mic92** [comment]
This works until people start mentioning you in more prs than you can read and you don't even get to see the other ones anymore.

**Comment by roberth** [comment]
You don't have to read it all to figure out what's relevant, but yeah, maybe I'm putting in too much effort.

**Comment by Lassulus** [comment]
I wanted to fix that but the TODO list was already to long. also I'm not super familiar with pkgs.formats (still wondering why it's in pkgs and not lib?) so it would probably have taken more than 15mins for me. thanks for fixing though

**Comment by zimbatm** [comment]
FYI: I saw it and decided to ignore it because of how emotionally charged the messages were. I just don't have the capacity to deal with more drama right now.

**Comment by roberth** [comment]
Next time you can tell me. Didn't mean to make drama.

**Pull request merged**